File: amqnullmessages/src/main/scala/codemettle/TestRepro.scala
package codemettle

import com.typesafe.scalalogging.StrictLogging
import org.springframework.beans.factory.{DisposableBean, InitializingBean}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration.Duration
import scala.concurrent.{Await, Future}

class TestRepro extends StrictLogging with InitializingBean with DisposableBean {
  private var systems = Seq.empty[HasSystem]

  override def afterPropertiesSet(): Unit = {
    logger.info("Initializing")

    // create listeners on TCP transports to simulate other JVMs consuming
    systems ++= (for (i <- 1 to 3)
      yield new TopicListeners(tcpActSys(s"Remote$i"), logNulls = false))

    // create listeners on VM transports to duplicate listeners on routers in AMQ
    systems ++= (for (i <- 1 to 3)
      yield new TopicListeners(vmActSys(s"Local$i"), logNulls = i == 1))

    // setup queue on VM transport that responds to messages
    systems :+= new QueueReplierSystem(vmActSys("Replier"))

    // create splitter on VM transport that takes in messages, sends them to queue, splits the messages, and sends them
    // to various topics to be consumed by TopicListeners
    systems :+= new Splitter(vmActSys("Splitter"))

    // create sender on TCP transport simulate external JVM process sending items to the splitter
    systems :+= new OriginalProducer(tcpActSys("Producer"))

    logger.info("Initialization done")
  }

  override def destroy(): Unit = {
    logger.info("Shutting down")

    val termsF = Future.sequence(systems.map(_.system.terminate()))
    Await.ready(termsF, Duration.Inf)

    logger.info("Completed shutting down")
  }
}

File: amqnullmessages/src/main/java/com/codemettle/TestRepro.java
package com.codemettle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.DisposableBean;
import org.springframework.beans.factory.InitializingBean;

import java.io.Closeable;
import java.io.IOException;
import java.util.LinkedList;
import java.util.List;

import static com.codemettle.Util.tcpSession;
import static com.codemettle.Util.vmSession;

@SuppressWarnings("unused")
public class TestRepro implements InitializingBean, DisposableBean {
    private Logger m_logger = LoggerFactory.getLogger(getClass());
    private List<Closeable> m_toClose = new LinkedList<>();

    @Override
    public void afterPropertiesSet() throws Exception {
        m_logger.info("Initializing");

        // create listeners on TCP transports to simulate other JVMs consuming
        for (int i = 0; i < 3; ++i)
            m_toClose.add(new TopicListeners(tcpSession(), false));

        // create listeners on VM transports to duplicate listeners on routers in AMQ
        for (int i = 0; i < 3; ++i)
            m_toClose.add(new TopicListeners(vmSession(), true));

        // create splitter on VM transport that takes in messages, sends them to queue, splits the messages, and sends
        // them to various topics to be consumed by TopicListeners
        m_toClose.add(new Splitter(vmSession()));

        // create sender on TCP transport simulate external JVM process sending items to the splitter
        m_toClose.add(new OriginalProducer(tcpSession()));

        m_logger.info("Initialization done");
    }

    private void shutdown(final Closeable aToClose) {
        try {
            aToClose.close();
        } catch (IOException e) {
            // ignore
        }
    }

    @SuppressWarnings("RedundantThrows")
    @Override
    public void destroy() throws Exception {
        m_logger.info("Shutting down");

        m_toClose.forEach(this::shutdown);

        m_logger.info("Completed shutting down");
    }
}
